shared_model/utils/amount_utils.hpp

```diff
@@ -18,62 +18,92 @@
 #ifndef IROHA_AMOUNT_UTILS_HPP
 #define IROHA_AMOUNT_UTILS_HPP
 
+#include "builders/default_builders.hpp"
 #include "builders/protobuf/common_objects/proto_amount_builder.hpp"
+#include "common/result.hpp"
+
+int pow(int degree, int number) {
+  return static_cast<int>(std::pow(degree, number));
+}
 
 /**
  * Sums up two amounts.
- * Requires to have the same scale.
- * Otherwise nullopt is returned
+ * Result is returned
  * @param a left term
  * @param b right term
  */
-iroha::expected::Result<std::shared_ptr<shared_model::interface::Amount>,
-                        std::shared_ptr<std::string>>
+iroha::expected::PolymorphicResult<shared_model::interface::Amount,
+                        std::string>
 operator+(const shared_model::interface::Amount &a,
           const shared_model::interface::Amount &b) {
-  // check precisions
-  if (a.precision() != b.precision()) {
-    return iroha::expected::makeError(
-        std::make_shared<std::string>("precision mismatch"));
-  }
-  if (a.intValue() + b.intValue() < a.intValue()
-      || a.intValue() + b.intValue() < b.intValue()) {
+  auto max_precision = std::max(a.precision(), b.precision());
```

Comment:
Consider making the code block which computes values for the same precisions into a separate function, since you do the same thing below